src/Sarif/Readers/JsonPositionedTextReader.cs

```diff
@@ -0,0 +1,44 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+using System;
+using System.IO;
+using Newtonsoft.Json;
+
+namespace Microsoft.CodeAnalysis.Sarif.Readers
+{
+    /// <summary>
+    ///  JsonPositionedTextReader is a JsonTextReader which exposes TokenPosition(),
+    ///  the byte offset in the stream where the current token begins. It is used to
+    ///  enable deferred collections, which look like other collections but are deserialized
+    ///  only when enumerated, saving memory.
+    /// </summary>
+    public class JsonPositionedTextReader : JsonTextReader
+    {
+        private LineMappingStreamReader StreamReader { get; set; }
+        private int LastReadLineNumber { get; set; }
+        private int LastReadLinePosition { get; set; }
+
+        public Func<Stream> StreamProvider { get; private set; }
```

Comment:
>StreamProvider [](start = 28, length = 14)

i'm not sure I understand the utility of this. in your usage, this delegate simply provides a reference to a stream, so it could just as easily have been passed as a ctor arg. if you did, in fact, use the delegate as some kind of factory, one confusion is how that stream is tracked for dispose, this type retains no reference to it.

Reply:
Added a comment on the member and constructor to explain. It's subtle.